replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/HeartbeaterAndVerifierImpl.java

```diff
@@ -73,23 +135,307 @@ public void cleanUp() {
+        LOGGER.trace("Heartbeat protocol not supported for site {}. Making REST head call.", url);
+        if (restSupported(url)) {
+          site.setVerifiedUrl(url);
+          site.setRemoteManaged(false);
+          siteManager.save(site);
+        }
+      }
+    } else {
+      LOGGER.trace("Doing nothing for already verified site {}", site.getVerifiedUrl());
+    }
   }
 
   @Override
   public void heartbeat(ReplicationSite site) throws InterruptedException {
-    // do nothing for now
-    // schedule the heartbeat to occur on another thread
-    // if the heartbeat fails, get all configurations that involves that site
-    //   and mark their status to Status.CONNECTION_LOST if the current status was XXXX_IN_PROGRESS
-    //   and to Status.CONNECTION_UNAVAILABLE for any other status
-    // if the heartbeat succeeds, get all configurations that involves that site
-    //   and if their current status is Status.CONNECTION_LOST or Status.CONNECTION_UNAVAILABLE,
-    //   mark their status to Status.PENDING indicating the remote end should restart soon to
-    //   replicate. For all other status, leave untouched
+    // it is entirely possible that we still end up with duplicates since we don't prevent another
+    // thread from checking and adding the same site to the queue. However, the probability is low
+    // of that happening and if it ever happened, will just end up send 2 successive heartbeats in
+    // a row
+    if (!pending.contains(site)) {
+      pending.put(site);
+    }
+  }
+
+  private boolean restSupported(String url) {
+    // use a generic WebClient because ddf RESTService interface does not have head, but the
+    // implementation supports it
+    SecureCxfClientFactory<WebClient> restClientFactory =
+        clientFactoryFactory.getSecureCxfClientFactory(
+            url + "/" + RESTService.CONTEXT_ROOT, WebClient.class);
+
+    WebClient restService = restClientFactory.getClientForSubject(security.getSystemSubject());
+
+    try {
+      return restService.head().getStatusInfo().getFamily().equals(Family.SUCCESSFUL);
+    } catch (ProcessingException e) {
+      LOGGER.debug("Error sending head to rest endpoint for site {}", url, e);
+    }
+
+    return false;
+  }
+
+  private HeartbeatResult doDiscovery(final String url) {
+    DiscoveryApi discovery = createDiscoveryClient(url);
+
+    ContactInfo contactInfo = new ContactInfo().email(SystemInfo.getSiteContatct());
+    com.connexta.discovery.rest.models.SystemInfo systemInfo =
+        new com.connexta.discovery.rest.models.SystemInfo()
+            .contact(contactInfo)
+            .id(System.getProperty("org.codice.ddf.system.registry-id"))
+            .name(SystemInfo.getSiteName())
+            .organization(SystemInfo.getOrganization())
+            .version(SystemInfo.getVersion())
+            .url(SystemBaseUrl.EXTERNAL.getBaseUrl())
+            .product(brandingPlugin.getProductName());
```

Comment:
Why do we need the branding plugin for this? Can we just use the system property `org.codice.ddf.system.branding`? I would like to avoid pulling in more ddf dependencies if possible.

Reply:
I went through the branding plugin cause it was what I was familiar with. We can go to the system property, yes. There is a small difference though. By default the plugin gives something like `DDF 1.2.3` but that system property will give `ddf` by default. 

It doesn't matter if we add more DDF dependencies here though cause this PR is against 0.2.x and will not be ported to master.